src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Subscriptions/DependencyRulesSubscriber.cs

```diff
@@ -184,12 +188,18 @@ private IReadOnlyCollection<string> GetWatchedRules(RuleHandlerType handlerType)
                 return;
             }
 
-            await _tasksService.LoadedProjectAsync(async () =>
+            // Ensure updates don't overlap
+            await _updateGate.ExecuteWithinLockAsync(JoinableCollection, JoinableFactory, async () =>
             {
-                using (ProjectCapabilitiesContext.CreateIsolatedContext(configuredProject, capabilities))
+                // Ensure the project doesn't unload during the update
+                await _tasksService.LoadedProjectAsync(async () =>
```

Comment:
Do we actually want to block project unload on this or just cancel this operation if the project is unloaded?

Reply:
Yes, very likely it is not resilient against unload - though probably want to eventually pass through CancellationToken to bail earlier.

Reply:
Added a TODO in fa9b08210.